### PR TITLE
[flang] Cray pointer in module

### PR DESCRIPTION
**Pull request by kkwli**
This patch is to add the support of declaring a Cray pointer in a module.

### CODE REVIEW

File: flang/lib/Semantics/mod-file.cpp [hunk 1]
```diff
@@ -759,6 +759,15 @@ void ModFileWriter::PutObjectEntity(
...
   }
+  if (symbol.test(Fortran::semantics::Symbol::Flag::CrayPointer)) {
+    if (!symbol.owner().crayPointers().empty()) {
+      for (const auto &[pointee, pointer] : symbol.owner().crayPointers()) {
+        if (pointer == symbol) {
+          os << "pointer(" << symbol.name() << "," << pointee << ")\n";
+        }
+      }
+    }
+  }
```
Comment (clementval) [inline flang/lib/Semantics/mod-file.cpp:770]: Can you add a test that check the mod file is correct?

File: flang/lib/Semantics/mod-file.cpp [hunk 2]
```diff
@@ -759,6 +759,15 @@ void ModFileWriter::PutObjectEntity(
     PutLower(os << "attributes(", common::EnumToString(*attr))
         << ") " << symbol.name() << '\n';
   }
+  if (symbol.test(Fortran::semantics::Symbol::Flag::CrayPointer)) {
```
Comment (vzakhari) [inline flang/lib/Semantics/mod-file.cpp:762]: Please add a LIT test with `test_modfile.py` in `flang/test/Semantics/`.

**Review by vzakhari — COMMENTED** [review]
Thank you! It looks good, but please add a semantics test.

File: flang/lib/Semantics/mod-file.cpp (hunk 1 above)
Comment (kkwli) [reply flang/lib/Semantics/mod-file.cpp:770]: Thanks for the review. Sure, I will do that.

File: flang/lib/Semantics/mod-file.cpp (hunk 2 above)
Comment (kkwli) [reply flang/lib/Semantics/mod-file.cpp:762]: Thanks for the review. I will do that.

**Comment by kkwli** [comment]
@vzakhari @clementval I added `flang/test/Semantics/modfile57.f90`.

File: flang/test/Semantics/modfile57.f90 [hunk 3]
```diff
@@ -0,0 +1,14 @@
+! RUN: %python %S/test_modfile.py %s %flang_fc1
+
+! Cray pointer
+module m
+  integer :: pte
+  pointer(ptr,pte)
```
Comment (vzakhari) [inline flang/test/Semantics/modfile57.f90:6]: Thank you for the test.  Can you please also add a case where the pointee's shape is defined by the `pointer` statement?

Something like this:
```
integer :: x
pointer(ptrx, x[10])
```

Comment (kkwli) [reply flang/test/Semantics/modfile57.f90:6]: Added array

**Review by vzakhari — APPROVED** [review]
Thank you! LGTM

**Comment by kiranchandramohan** [comment]
Thanks @kkwli.

Can we enable the Cray pointer tests in the gfortran-testuite by removing it from the Disabled list (https://github.com/llvm/llvm-test-suite/blob/main/Fortran/gfortran/regression/DisabledFiles.cmake)?

**Comment by kkwli** [comment]
> Thanks @kkwli.
> 
> Can we enable the Cray pointer tests in the gfortran-testuite by removing it from the Disabled list (https://github.com/llvm/llvm-test-suite/blob/main/Fortran/gfortran/regression/DisabledFiles.cmake)?

@kiranchandramohan Sure, I will do that.

**Pull request merged**